File: src/hswm/cells/adaptive_store.py
"""Small durable store for versioned, role-typed adaptive hypergraph atoms.

It is a local prototype persistence boundary.  It records immutable revisions
and compare-and-swap rewrites; it neither decides learning efficacy nor writes
to the reference KG.
"""

from __future__ import annotations

from contextlib import contextmanager
from hashlib import sha256
import json
from pathlib import Path
import sqlite3
from typing import Any


SCHEMA_VERSION = "hswm-adaptive-hypergraph/v1"


class AdaptiveStoreError(RuntimeError):
    pass


def _bytes(value: Any) -> bytes:
    try:
        return json.dumps(value, ensure_ascii=False, allow_nan=False, sort_keys=True,
                          separators=(",", ":")).encode()
    except (TypeError, ValueError) as error:
        raise AdaptiveStoreError("value is not canonical JSON") from error


def _digest(value: Any) -> str:
    return sha256(_bytes(value)).hexdigest()


def _load(raw: bytes, label: str) -> Any:
    try:
        value = json.loads(raw)
    except json.JSONDecodeError as error:
        raise AdaptiveStoreError(f"stored {label} is malformed") from error
    if _bytes(value) != raw:
        raise AdaptiveStoreError(f"stored {label} is not canonical")
    return value


def _atom(value: Any) -> dict[str, Any]:
    if not isinstance(value, dict) or set(value) != {"uid", "kind", "owner", "refs", "payload"}:
        raise AdaptiveStoreError("atom shape is invalid")
    if any(not isinstance(value[key], str) or not value[key] for key in ("uid", "kind", "owner")):
        raise AdaptiveStoreError("atom identity is invalid")
    if not isinstance(value["refs"], list):
        raise AdaptiveStoreError("atom refs are invalid")
    refs: list[dict[str, str]] = []
    seen: set[tuple[str, str]] = set()
    for ref in value["refs"]:
        if (not isinstance(ref, dict) or set(ref) != {"role", "uid"} or
                not isinstance(ref["role"], str) or not ref["role"] or
                not isinstance(ref["uid"], str) or not ref["uid"]):
            raise AdaptiveStoreError("atom reference is invalid")
        key = (ref["role"], ref["uid"])
        if key in seen:
            raise AdaptiveStoreError("duplicate role reference")
        seen.add(key)
        refs.append({"role": ref["role"], "uid": ref["uid"]})
    _bytes(value["payload"])
    return {"uid": value["uid"], "kind": value["kind"], "owner": value["owner"],
            "refs": refs, "payload": value["payload"]}


class AdaptiveStore:
    """SQLite CAS store for one or more locally named adaptive graphs."""

    def __init__(self, path: str | Path) -> None:
        self.path = Path(path)
        self.path.parent.mkdir(parents=True, exist_ok=True)
        self._initialize()

    def _connect(self) -> sqlite3.Connection:
        connection = sqlite3.connect(self.path, isolation_level=None, timeout=10)
        connection.row_factory = sqlite3.Row
        connection.execute("PRAGMA foreign_keys=ON")
        return connection

    @contextmanager
    def _connection(self):
        connection = self._connect()
        try:
            yield connection
        finally:
            connection.close()

    def _initialize(self) -> None:
        with self._connection() as connection:
            connection.execute("PRAGMA journal_mode=WAL")
            connection.executescript("""
            CREATE TABLE IF NOT EXISTS adaptive_graphs (
              graph_id TEXT PRIMARY KEY, manifest_digest TEXT NOT NULL
            );
            CREATE TABLE IF NOT EXISTS adaptive_atoms (
              graph_id TEXT NOT NULL, uid TEXT NOT NULL, revision INTEGER NOT NULL,
              kind TEXT NOT NULL, owner TEXT NOT NULL, refs_json BLOB NOT NULL,
              payload_json BLOB NOT NULL, atom_digest TEXT NOT NULL, event_id TEXT,
              PRIMARY KEY (graph_id, uid, revision),
              FOREIGN KEY (graph_id) REFERENCES adaptive_graphs(graph_id)
            );
            CREATE TABLE IF NOT EXISTS adaptive_heads (
              graph_id TEXT NOT NULL, uid TEXT NOT NULL, revision INTEGER NOT NULL,
              kind TEXT NOT NULL, owner TEXT NOT NULL, atom_digest TEXT NOT NULL,
              PRIMARY KEY (graph_id, uid),
              FOREIGN KEY (graph_id, uid, revision) REFERENCES adaptive_atoms(graph_id, uid, revision)
            );
            CREATE TABLE IF NOT EXISTS adaptive_events (
              graph_id TEXT NOT NULL, event_id TEXT NOT NULL, intent_digest TEXT NOT NULL,
              source_json BLOB NOT NULL, consumed_json BLOB NOT NULL, produced_json BLOB NOT NULL,
              PRIMARY KEY (graph_id, event_id),
              FOREIGN KEY (graph_id) REFERENCES adaptive_graphs(graph_id)
            );
            """)

    def initialize(self, graph_id: str, manifest_digest: str, atoms: list[dict[str, Any]]) -> list[dict[str, Any]]:
        if not isinstance(graph_id, str) or not graph_id or not isinstance(manifest_digest, str) or not manifest_digest:
            raise AdaptiveStoreError("graph identity is invalid")
        rows = [_atom(item) for item in atoms]
        if not rows or len({row["uid"] for row in rows}) != len(rows):
            raise AdaptiveStoreError("initial atoms must be unique and non-empty")
        uids = {row["uid"] for row in rows}
        if any(ref["uid"] not in uids for row in rows for ref in row["refs"]):
            raise AdaptiveStoreError("initial reference is unresolved")
        with self._connection() as connection:
            connection.execute("BEGIN IMMEDIATE")
            existing = connection.execute("SELECT manifest_digest FROM adaptive_graphs WHERE graph_id=?", (graph_id,)).fetchone()
            if existing is not None:
                if existing["manifest_digest"] != manifest_digest:
                    connection.rollback()
                    raise AdaptiveStoreError("graph manifest digest conflict")
                connection.commit()
                return self.heads(graph_id)
            connection.execute("INSERT INTO adaptive_graphs VALUES(?,?)", (graph_id, manifest_digest))
            for row in rows:
                atom_digest = _digest(row)
                connection.execute("INSERT INTO adaptive_atoms VALUES(?,?,?,?,?,?,?,?,?)", (
                    graph_id, row["uid"], 1, row["kind"], row["owner"], _bytes(row["refs"]),
                    _bytes(row["payload"]), atom_digest, None))
                connection.execute("INSERT INTO adaptive_heads VALUES(?,?,?,?,?,?)", (
                    graph_id, row["uid"], 1, row["kind"], row["owner"], atom_digest))
            connection.commit()
        return self.heads(graph_id)

    def _graph(self, connection: sqlite3.Connection, graph_id: str) -> None:
        if connection.execute("SELECT 1 FROM adaptive_graphs WHERE graph_id=?", (graph_id,)).fetchone() is None:
            raise AdaptiveStoreError("graph does not exist")

    @staticmethod
    def _head(row: sqlite3.Row) -> dict[str, Any]:
        return {key: row[key] for key in ("uid", "revision", "kind", "owner", "digest")}

    def head(self, graph_id: str, uid: str) -> dict[str, Any]:
        with self._connection() as connection:
            self._graph(connection, graph_id)
            row = connection.execute("SELECT uid,revision,kind,owner,atom_digest AS digest FROM adaptive_heads WHERE graph_id=? AND uid=?", (graph_id, uid)).fetchone()
            if row is None:
                raise AdaptiveStoreError("atom head does not exist")
            return self._head(row)

    def heads(self, graph_id: str, kind: str | None = None) -> list[dict[str, Any]]:
        with self._connection() as connection:
            self._graph(connection, graph_id)
            query = "SELECT uid,revision,kind,owner,atom_digest AS digest FROM adaptive_heads WHERE graph_id=?"
            values: tuple[Any, ...] = (graph_id,)
            if kind is not None:
                query += " AND kind=?"
                values += (kind,)
            return [self._head(row) for row in connection.execute(query + " ORDER BY uid", values)]

    def get_revision(self, graph_id: str, uid: str, revision: int) -> dict[str, Any]:
        if type(revision) is not int or revision < 1:
            raise AdaptiveStoreError("atom revision is invalid")
        with self._connection() as connection:
            self._graph(connection, graph_id)
            row = connection.execute("SELECT * FROM adaptive_atoms WHERE graph_id=? AND uid=? AND revision=?", (graph_id, uid, revision)).fetchone()
            if row is None:
                raise AdaptiveStoreError("atom revision does not exist")
            atom = _atom({"uid": row["uid"], "kind": row["kind"], "owner": row["owner"],
                          "refs": _load(bytes(row["refs_json"]), "refs"),
                          "payload": _load(bytes(row["payload_json"]), "payload")})
            if _digest(atom) != row["atom_digest"]:
                raise AdaptiveStoreError("stored atom digest mismatch")
            return {**atom, "revision": row["revision"], "digest": row["atom_digest"], "event_id": row["event_id"]}

    def rewrite(self, graph_id: str, *, event_id: str, expected: dict[str, int], atoms: list[dict[str, Any]], source: dict[str, Any]) -> dict[str, Any]:
        if not isinstance(event_id, str) or not event_id or not isinstance(expected, dict) or not isinstance(source, dict):
            raise AdaptiveStoreError("rewrite identity is invalid")
        if any(not isinstance(uid, str) or not uid or type(revision) is not int or revision < 0 for uid, revision in expected.items()):
            raise AdaptiveStoreError("rewrite expectations are invalid")
        rows = [_atom(item) for item in atoms]
        if not rows or len({row["uid"] for row in rows}) != len(rows) or set(expected) != {row["uid"] for row in rows}:
            raise AdaptiveStoreError("rewrite atoms must exactly match expectations")
        intent = {"expected": expected, "atoms": rows, "source": source}
        intent_digest = _digest(intent)
        with self._connection() as connection:
            connection.execute("BEGIN IMMEDIATE")
            self._graph(connection, graph_id)
            prior = connection.execute("SELECT * FROM adaptive_events WHERE graph_id=? AND event_id=?", (graph_id, event_id)).fetchone()
            if prior is not None:
                if prior["intent_digest"] != intent_digest:
                    connection.rollback()
                    raise AdaptiveStoreError("event id was reused with a different intent")
                connection.commit()
                return self.get_event(graph_id, event_id)
            current = {row["uid"]: row for row in connection.execute("SELECT * FROM adaptive_heads WHERE graph_id=?", (graph_id,))}
            transaction_uids = {row["uid"] for row in rows}
            for row in rows:
                prior_head = current.get(row["uid"])
                wanted = expected[row["uid"]]
                if prior_head is None:
                    if wanted != 0:
                        connection.rollback()
                        raise AdaptiveStoreError("new atom must expect revision zero")
                elif prior_head["revision"] != wanted:
                    connection.rollback()
                    raise AdaptiveStoreError("head compare-and-swap conflict")
                elif prior_head["owner"] != row["owner"]:
                    connection.rollback()
                    raise AdaptiveStoreError("atom owner is immutable")
                elif prior_head["kind"] != row["kind"]:
                    connection.rollback()
                    raise AdaptiveStoreError("atom kind is immutable")
                for ref in row["refs"]:
                    if ref["uid"] not in current and ref["uid"] not in transaction_uids:
                        connection.rollback()
                        raise AdaptiveStoreError("rewrite reference is unresolved")
            consumed = [
                {"uid": row["uid"], "revision": row["revision"], "kind": row["kind"],
                 "owner": row["owner"], "digest": row["atom_digest"]}
                for row in current.values() if row["uid"] in expected
            ]
            produced: list[dict[str, Any]] = []
            for row in rows:
                prior_head = current.get(row["uid"])
                revision = 1 if prior_head is None else prior_head["revision"] + 1
                atom_digest = _digest(row)
                connection.execute("INSERT INTO adaptive_atoms VALUES(?,?,?,?,?,?,?,?,?)", (graph_id, row["uid"], revision, row["kind"], row["owner"], _bytes(row["refs"]), _bytes(row["payload"]), atom_digest, event_id))
                if prior_head is None:
                    connection.execute("INSERT INTO adaptive_heads VALUES(?,?,?,?,?,?)", (graph_id, row["uid"], revision, row["kind"], row["owner"], atom_digest))
                else:
                    connection.execute("UPDATE adaptive_heads SET revision=?,kind=?,owner=?,atom_digest=? WHERE graph_id=? AND uid=?", (revision, row["kind"], row["owner"], atom_digest, graph_id, row["uid"]))
                produced.append({"uid": row["uid"], "revision": revision, "kind": row["kind"], "owner": row["owner"], "digest": atom_digest})
            connection.execute("INSERT INTO adaptive_events VALUES(?,?,?,?,?,?)", (graph_id, event_id, intent_digest, _bytes(source), _bytes(consumed), _bytes(produced)))
            connection.commit()
        return {"event_id": event_id, "intent_digest": intent_digest, "consumed": consumed,
                "produced": produced, "output_heads": produced}

    def get_event(self, graph_id: str, event_id: str) -> dict[str, Any]:
        with self._connection() as connection:
            self._graph(connection, graph_id)
            row = connection.execute("SELECT * FROM adaptive_events WHERE graph_id=? AND event_id=?", (graph_id, event_id)).fetchone()
            if row is None:
                raise AdaptiveStoreError("event does not exist")
            produced = _load(bytes(row["produced_json"]), "produced")
            return {"event_id": event_id, "intent_digest": row["intent_digest"],
                    "source": _load(bytes(row["source_json"]), "source"),
                    "consumed": _load(bytes(row["consumed_json"]), "consumed"),
                    "produced": produced, "output_heads": produced}

    def events(self, graph_id: str) -> list[dict[str, Any]]:
        with self._connection() as connection:
            self._graph(connection, graph_id)
            ids = [row["event_id"] for row in connection.execute("SELECT event_id FROM adaptive_events WHERE graph_id=? ORDER BY rowid", (graph_id,))]
        return [self.get_event(graph_id, event_id) for event_id in ids]

File: src/hswm/effect-runtime/src/adaptive-store.ts
/**
 * Durable, trusted-local persistence for the adaptive prototype. This is not a
 * Canonical Atom v2 admission path.
 */
import { createHash, randomUUID } from "node:crypto";
import { mkdirSync } from "node:fs";
import { dirname } from "node:path";
import { DatabaseSync } from "node:sqlite";
import { Context, Data, Effect, Either, Layer } from "effect";
export interface AdaptiveAtomReference {
    readonly role: string;
    readonly uid: string;
}
export interface AdaptiveAtom {
    readonly uid: string;
    readonly kind: string;
    readonly owner: string;
    readonly refs: ReadonlyArray<AdaptiveAtomReference>;
    readonly payload: unknown;
}
export interface AdaptiveAtomHead {
    readonly uid: string;
    readonly revision: number;
    readonly kind: string;
    readonly owner: string;
    readonly digest: string;
}
export interface AdaptiveAtomRevision extends AdaptiveAtom {
    readonly revision: number;
    readonly digest: string;
    readonly eventId: string | null;
}
export interface AdaptiveStoreEvent {
    readonly eventId: string;
    readonly intentDigest: string;
    readonly source: Record<string, unknown>;
    readonly consumed: ReadonlyArray<AdaptiveAtomHead>;
    readonly produced: ReadonlyArray<AdaptiveAtomHead>;
    readonly outputHeads: ReadonlyArray<AdaptiveAtomHead>;
}
export interface AdaptiveRewrite {
    readonly graphId: string;
    readonly eventId: string;
    readonly expected: Readonly<Record<string, number>>;
    readonly atoms: ReadonlyArray<AdaptiveAtom>;
    readonly source: Readonly<Record<string, unknown>>;
}
export class AdaptiveStoreError extends Data.TaggedError("AdaptiveStoreError")<{
    readonly operation: "INITIALIZE" | "HEAD" | "HEADS" | "GET_REVISION" | "REWRITE" | "GET_EVENT" | "EVENTS" | "ACQUIRE_RUNTIME_LOCK" | "RELEASE_RUNTIME_LOCK" | "CLOSE";
    readonly detail: string;
}> {
}
export class AdaptiveStore extends Context.Tag("hswm/AdaptiveStore")<AdaptiveStore, {
    readonly initialize: (graphId: string, manifestDigest: string, atoms: ReadonlyArray<AdaptiveAtom>) => Effect.Effect<ReadonlyArray<AdaptiveAtomHead>, AdaptiveStoreError>;
    readonly head: (graphId: string, uid: string) => Effect.Effect<AdaptiveAtomHead, AdaptiveStoreError>;
    readonly heads: (graphId: string, kind?: string) => Effect.Effect<ReadonlyArray<AdaptiveAtomHead>, AdaptiveStoreError>;
    readonly getRevision: (graphId: string, uid: string, revision: number) => Effect.Effect<AdaptiveAtomRevision, AdaptiveStoreError>;
    readonly rewrite: (input: AdaptiveRewrite) => Effect.Effect<AdaptiveStoreEvent, AdaptiveStoreError>;
    readonly getEvent: (graphId: string, eventId: string) => Effect.Effect<AdaptiveStoreEvent, AdaptiveStoreError>;
    readonly events: (graphId: string) => Effect.Effect<ReadonlyArray<AdaptiveStoreEvent>, AdaptiveStoreError>;
    readonly acquireRuntimeLock: (graphId: string) => Effect.Effect<string, AdaptiveStoreError>;
    readonly releaseRuntimeLock: (graphId: string, token: string) => Effect.Effect<void, AdaptiveStoreError>;
}>() {
}
type Op = AdaptiveStoreError["operation"];
type Row = Record<string, unknown>;
type Result<A> = Either.Either<A, AdaptiveStoreError>;
const fail = <A = never>(operation: Op, detail: string): Result<A> => Either.left(new AdaptiveStoreError({ operation, detail }));
const message = (cause: unknown, fallback: string) => cause instanceof Error ? cause.message : fallback;
const db = <A>(operation: Op, action: () => A): Result<A> => Either.try({ try: action, catch: (cause) => new AdaptiveStoreError({ operation, detail: message(cause, "SQLite operation failed") }) });
const toEffect = <A>(result: Result<A>): Effect.Effect<A, AdaptiveStoreError> => Either.match(result, { onLeft: Effect.fail, onRight: Effect.succeed });
const record = (value: unknown): value is Record<string, unknown> => typeof value === "object" && value !== null && !Array.isArray(value);
const hash = (value: string) => createHash("sha256").update(value, "utf8").digest("hex");
const row = (value: unknown) => value as Row | undefined;
const rows = (value: unknown) => value as ReadonlyArray<Row>;
const string = (value: unknown) => String(value);
const numeric = (value: unknown) => typeof value === "number" ? value : Number(value);
const head = (value: Row): AdaptiveAtomHead => Object.freeze({ uid: string(value["uid"]), revision: numeric(value["revision"]), kind: string(value["kind"]), owner: string(value["owner"]), digest: string(value["digest"] ?? value["atom_digest"]) });
const blob = (value: unknown) => Buffer.from(value as Uint8Array).toString("utf8");
const json = (value: unknown, operation: Op, depth = 0): Result<string> => {
    if (depth > 64)
        return fail(operation, "JSON nesting exceeds the bounded store surface");
    if (value === null || typeof value === "string" || typeof value === "boolean")
        return Either.right(JSON.stringify(value));
    if (typeof value === "number")
        return Number.isFinite(value) ? Either.right(JSON.stringify(value)) : fail(operation, "value is not canonical JSON");
    if (Array.isArray(value))
        return Either.all(value.map((item) => json(item, operation, depth + 1))).pipe(Either.map((items) => `[${items.join(",")}]`));
    if (!record(value))
        return fail(operation, "value is not canonical JSON");
    return Either.all(Object.keys(value).sort().map((key) => json(value[key], operation, depth + 1).pipe(Either.map((item) => `${JSON.stringify(key)}:${item}`)))).pipe(Either.map((items) => `{${items.join(",")}}`));
};
const atom = (value: unknown, operation: Op): Result<AdaptiveAtom> => Either.gen(function* () {
    if (!record(value) || Object.keys(value).length !== 5 || !["uid", "kind", "owner", "refs", "payload"].every((key) => key in value))
        return yield* fail(operation, "atom shape is invalid");
    const { uid, kind, owner, refs, payload } = value;
    if (typeof uid !== "string" || !uid || typeof kind !== "string" || !kind || typeof owner !== "string" || !owner || !Array.isArray(refs))
        return yield* fail(operation, "atom identity or refs are invalid");
    const checked = yield* Either.all(refs.map((ref) => record(ref) && Object.keys(ref).length === 2 && typeof ref["role"] === "string" && !!ref["role"] && typeof ref["uid"] === "string" && !!ref["uid"] ? Either.right(Object.freeze({ role: ref["role"], uid: ref["uid"] })) : fail<AdaptiveAtomReference>(operation, "atom reference is invalid")));
    if (new Set(checked.map((ref) => `${ref["role"]}\u0000${ref["uid"]}`)).size !== checked.length)
        return yield* fail(operation, "duplicate role reference");
    yield* json(payload, operation);
    return Object.freeze({ uid, kind, owner, refs: Object.freeze(checked), payload });
});
// Stored Python JSON bytes are intentional: a historic 1.0 must not become 1.
const rawAtom = (uid: string, kind: string, owner: string, refs: string, payload: string) => `{\"kind\":${JSON.stringify(kind)},\"owner\":${JSON.stringify(owner)},\"payload\":${payload},\"refs\":${refs},\"uid\":${JSON.stringify(uid)}}`;
const digest = (value: AdaptiveAtom, operation: Op) => json(value, operation).pipe(Either.map(hash));
const schema = `CREATE TABLE IF NOT EXISTS adaptive_graphs (graph_id TEXT PRIMARY KEY, manifest_digest TEXT NOT NULL);
CREATE TABLE IF NOT EXISTS adaptive_atoms (graph_id TEXT NOT NULL, uid TEXT NOT NULL, revision INTEGER NOT NULL, kind TEXT NOT NULL, owner TEXT NOT NULL, refs_json BLOB NOT NULL, payload_json BLOB NOT NULL, atom_digest TEXT NOT NULL, event_id TEXT, PRIMARY KEY (graph_id, uid, revision), FOREIGN KEY (graph_id) REFERENCES adaptive_graphs(graph_id));
CREATE TABLE IF NOT EXISTS adaptive_heads (graph_id TEXT NOT NULL, uid TEXT NOT NULL, revision INTEGER NOT NULL, kind TEXT NOT NULL, owner TEXT NOT NULL, atom_digest TEXT NOT NULL, PRIMARY KEY (graph_id, uid), FOREIGN KEY (graph_id, uid, revision) REFERENCES adaptive_atoms(graph_id, uid, revision));
CREATE TABLE IF NOT EXISTS adaptive_events (graph_id TEXT NOT NULL, event_id TEXT NOT NULL, intent_digest TEXT NOT NULL, source_json BLOB NOT NULL, consumed_json BLOB NOT NULL, produced_json BLOB NOT NULL, PRIMARY KEY (graph_id, event_id), FOREIGN KEY (graph_id) REFERENCES adaptive_graphs(graph_id));
CREATE TABLE IF NOT EXISTS adaptive_runtime_locks (graph_id TEXT PRIMARY KEY, token TEXT NOT NULL, pid INTEGER NOT NULL, FOREIGN KEY (graph_id) REFERENCES adaptive_graphs(graph_id));`;
const openDatabase = (path: string): Effect.Effect<DatabaseSync, AdaptiveStoreError> => {
    if (!path)
        return Effect.fail(new AdaptiveStoreError({ operation: "INITIALIZE", detail: "store path is invalid" }));
    const directory: Effect.Effect<void, AdaptiveStoreError> = path === ":memory:" ? Effect.succeed(undefined) : Effect.try({ try: () => mkdirSync(dirname(path), { recursive: true }), catch: (cause) => new AdaptiveStoreError({ operation: "INITIALIZE", detail: message(cause, "store directory initialization failed") }) }).pipe(Effect.asVoid);
    return directory.pipe(Effect.flatMap(() => Effect.suspend(() => toEffect(db("INITIALIZE", () => new DatabaseSync(path, { timeout: 10000, allowExtension: false }))))));
};
export const makeAdaptiveStoreSqliteLayer = (path: string) => Layer.scoped(AdaptiveStore, Effect.acquireRelease(openDatabase(path), (database) => toEffect(db("CLOSE", () => database.close())).pipe(Effect.orDie)).pipe(Effect.tap((database) => Effect.suspend(() => toEffect(db("INITIALIZE", () => database.exec("PRAGMA foreign_keys=ON; PRAGMA journal_mode=WAL;"))))), Effect.tap((database) => Effect.suspend(() => toEffect(db("INITIALIZE", () => database.exec(schema))))), Effect.map((database) => AdaptiveStore.of(service(database)))));
const service = (database: DatabaseSync) => {
    const native = <A>(operation: Op, action: () => A) => db(operation, action);
    const graph = (operation: Op, graphId: string): Result<void> => native(operation, () => row(database.prepare("SELECT 1 FROM adaptive_graphs WHERE graph_id=?").get(graphId))).pipe(Either.flatMap((found) => found ? Either.right(undefined) : fail(operation, "graph does not exist")));
    const storedJson = (operation: Op, value: unknown, label: string): Result<unknown> => native(operation, () => row(database.prepare("SELECT json_valid(CAST(? AS TEXT)) AS valid").get(value as Uint8Array))).pipe(Either.flatMap((valid) => numeric(valid?.["valid"]) === 1 ? Either.right(JSON.parse(blob(value))) : fail(operation, `stored ${label} is malformed`)));
    const allHeads = (operation: Op, graphId: string, kind?: string): Result<ReadonlyArray<AdaptiveAtomHead>> => Either.gen(function* () {
        yield* graph(operation, graphId);
        const found = yield* native(operation, () => kind === undefined ? rows(database.prepare("SELECT uid,revision,kind,owner,atom_digest AS digest FROM adaptive_heads WHERE graph_id=? ORDER BY uid").all(graphId)) : rows(database.prepare("SELECT uid,revision,kind,owner,atom_digest AS digest FROM adaptive_heads WHERE graph_id=? AND kind=? ORDER BY uid").all(graphId, kind)));
        return Object.freeze(found.map(head));
    });
    const event = (operation: Op, graphId: string, eventId: string): Result<AdaptiveStoreEvent> => Either.gen(function* () {
        yield* graph(operation, graphId);
        const found = yield* native(operation, () => row(database.prepare("SELECT * FROM adaptive_events WHERE graph_id=? AND event_id=?").get(graphId, eventId)));
        if (!found)
            return yield* fail(operation, "event does not exist");
        const source = yield* storedJson(operation, found["source_json"], "source");
        const consumed = yield* storedJson(operation, found["consumed_json"], "consumed");
        const produced = yield* storedJson(operation, found["produced_json"], "produced");
        const parse = (items: unknown): Result<ReadonlyArray<AdaptiveAtomHead>> => Array.isArray(items) ? Either.all(items.map((item) => record(item) ? Either.right(head(item)) : fail<AdaptiveAtomHead>(operation, "stored event is malformed"))).pipe(Either.map((items) => Object.freeze(items))) : fail(operation, "stored event is malformed");
        if (!record(source))
            return yield* fail(operation, "stored event is malformed");
        const outputHeads = yield* parse(produced);
        return Object.freeze({ eventId, intentDigest: string(found["intent_digest"]), source, consumed: yield* parse(consumed), produced: outputHeads, outputHeads });
    });
    const transaction = <A>(operation: Op, body: () => Result<A>): Result<A> => Either.gen(function* () {
        yield* native(operation, () => database.exec("BEGIN IMMEDIATE"));
        const outcome = body();
        if (Either.isRight(outcome)) {
            const committed = native(operation, () => database.exec("COMMIT"));
            if (Either.isRight(committed))
                return outcome.right;
            native(operation, () => database.exec("ROLLBACK"));
            return yield* Either.map(committed, () => outcome.right);
        }
        native(operation, () => database.exec("ROLLBACK"));
        return yield* outcome;
    });
    const effect = <A>(evaluate: () => Result<A>): Effect.Effect<A, AdaptiveStoreError> => Effect.suspend(() => toEffect(evaluate()));
    const live = (pid: number) => { try {
        process.kill(pid, 0);
        return true;
    }
    catch (cause) {
        return !(record(cause) && cause["code"] === "ESRCH");
    } };
    return {
        initialize: (graphId: string, manifestDigest: string, input: ReadonlyArray<AdaptiveAtom>) => effect(() => Either.gen(function* () {
            if (!graphId || !manifestDigest)
                return yield* fail("INITIALIZE", "graph identity is invalid");
            const atoms = yield* Either.all(input.map((item) => atom(item, "INITIALIZE")));
            if (!atoms.length || new Set(atoms.map((item) => item["uid"])).size !== atoms.length)
                return yield* fail("INITIALIZE", "initial atoms must be unique and non-empty");
            const uids = new Set(atoms.map((item) => item["uid"]));
            if (atoms.some((item) => item.refs.some((ref) => !uids.has(ref["uid"]))))
                return yield* fail("INITIALIZE", "initial reference is unresolved");
            return yield* transaction("INITIALIZE", () => Either.gen(function* () {
                const old = yield* native("INITIALIZE", () => row(database.prepare("SELECT manifest_digest FROM adaptive_graphs WHERE graph_id=?").get(graphId)));
                if (old)
                    return string(old["manifest_digest"]) === manifestDigest ? yield* allHeads("INITIALIZE", graphId) : yield* fail("INITIALIZE", "graph manifest digest conflict");
                yield* native("INITIALIZE", () => database.prepare("INSERT INTO adaptive_graphs VALUES(?,?)").run(graphId, manifestDigest));
                const putAtom = yield* native("INITIALIZE", () => database.prepare("INSERT INTO adaptive_atoms VALUES(?,?,?,?,?,?,?,?,?)"));
                const putHead = yield* native("INITIALIZE", () => database.prepare("INSERT INTO adaptive_heads VALUES(?,?,?,?,?,?)"));
                for (const item of atoms) {
                    const refs = yield* json(item.refs, "INITIALIZE");
                    const payload = yield* json(item.payload, "INITIALIZE");
                    const itemDigest = yield* digest(item, "INITIALIZE");
                    yield* native("INITIALIZE", () => putAtom.run(graphId, item["uid"], 1, item["kind"], item["owner"], Buffer.from(refs), Buffer.from(payload), itemDigest, null));
                    yield* native("INITIALIZE", () => putHead.run(graphId, item["uid"], 1, item["kind"], item["owner"], itemDigest));
                }
                return yield* allHeads("INITIALIZE", graphId);
            }));
        })),
        head: (graphId: string, uid: string) => effect(() => Either.gen(function* () { yield* graph("HEAD", graphId); const found = yield* native("HEAD", () => row(database.prepare("SELECT uid,revision,kind,owner,atom_digest AS digest FROM adaptive_heads WHERE graph_id=? AND uid=?").get(graphId, uid))); return found ? head(found) : yield* fail("HEAD", "atom head does not exist"); })),
        heads: (graphId: string, kind?: string) => effect(() => allHeads("HEADS", graphId, kind)),
        getRevision: (graphId: string, uid: string, revision: number) => effect(() => Either.gen(function* () {
            if (!Number.isInteger(revision) || revision < 1)
                return yield* fail("GET_REVISION", "atom revision is invalid");
            yield* graph("GET_REVISION", graphId);
            const found = yield* native("GET_REVISION", () => row(database.prepare("SELECT * FROM adaptive_atoms WHERE graph_id=? AND uid=? AND revision=?").get(graphId, uid, revision)));
            if (!found)
                return yield* fail("GET_REVISION", "atom revision does not exist");
            const refsRaw = blob(found["refs_json"]);
            const payloadRaw = blob(found["payload_json"]);
            const refs = yield* storedJson("GET_REVISION", found["refs_json"], "refs");
            const payload = yield* storedJson("GET_REVISION", found["payload_json"], "payload");
            const item = yield* atom({ uid: string(found["uid"]), kind: string(found["kind"]), owner: string(found["owner"]), refs, payload }, "GET_REVISION");
            if (hash(rawAtom(item["uid"], item["kind"], item["owner"], refsRaw, payloadRaw)) !== string(found["atom_digest"]))
                return yield* fail("GET_REVISION", "stored atom digest mismatch");
            return Object.freeze({ ...item, revision: numeric(found["revision"]), digest: string(found["atom_digest"]), eventId: found["event_id"] === null ? null : string(found["event_id"]) });
        })),
        rewrite: (input: AdaptiveRewrite) => effect(() => Either.gen(function* () {
            if (!input.graphId || !input.eventId || !record(input.expected) || !record(input.source))
                return yield* fail("REWRITE", "rewrite identity is invalid");
            if (Object.entries(input.expected).some(([uid, revision]) => !uid || !Number.isInteger(revision) || revision < 0))
                return yield* fail("REWRITE", "rewrite expectations are invalid");
            const atoms = yield* Either.all(input.atoms.map((item) => atom(item, "REWRITE")));
            if (!atoms.length || new Set(atoms.map((item) => item["uid"])).size !== atoms.length || Object.keys(input.expected).length !== atoms.length || atoms.some((item) => !(item["uid"] in input.expected)))
                return yield* fail("REWRITE", "rewrite atoms must exactly match expectations");
            const intent = hash(yield* json({ expected: input.expected, atoms, source: input.source }, "REWRITE"));
            return yield* transaction("REWRITE", () => Either.gen(function* () {
                yield* graph("REWRITE", input.graphId);
                const previousEvent = yield* native("REWRITE", () => row(database.prepare("SELECT intent_digest FROM adaptive_events WHERE graph_id=? AND event_id=?").get(input.graphId, input.eventId)));
                if (previousEvent)
                    return string(previousEvent["intent_digest"]) === intent ? yield* event("REWRITE", input.graphId, input.eventId) : yield* fail("REWRITE", "event id was reused with a different intent");
                const currentRows = yield* native("REWRITE", () => rows(database.prepare("SELECT * FROM adaptive_heads WHERE graph_id=?").all(input.graphId)));
                const current = new Map(currentRows.map((item) => [string(item["uid"]), item]));
                const touched = new Set(atoms.map((item) => item["uid"]));
                const valid = atoms.every((item) => { const old = current.get(item["uid"]); const expected = input.expected[item["uid"]]; return (!old ? expected === 0 : numeric(old["revision"]) === expected && string(old["owner"]) === item["owner"] && string(old["kind"]) === item["kind"]) && item.refs.every((ref) => current.has(ref["uid"]) || touched.has(ref["uid"])); });
                if (!valid)
                    return yield* fail("REWRITE", "head compare-and-swap conflict, immutable identity conflict, or unresolved reference");
                const consumed = Object.freeze(currentRows.filter((item) => string(item["uid"]) in input.expected).map(head));
                const produced: AdaptiveAtomHead[] = [];
                const putAtom = yield* native("REWRITE", () => database.prepare("INSERT INTO adaptive_atoms VALUES(?,?,?,?,?,?,?,?,?)"));
                const putHead = yield* native("REWRITE", () => database.prepare("INSERT INTO adaptive_heads VALUES(?,?,?,?,?,?)"));
                const updateHead = yield* native("REWRITE", () => database.prepare("UPDATE adaptive_heads SET revision=?,kind=?,owner=?,atom_digest=? WHERE graph_id=? AND uid=?"));
                for (const item of atoms) {
                    const old = current.get(item["uid"]);
                    const next = old ? numeric(old["revision"]) + 1 : 1;
                    const itemDigest = yield* digest(item, "REWRITE");
                    const refs = yield* json(item.refs, "REWRITE");
                    const payload = yield* json(item.payload, "REWRITE");
                    yield* native("REWRITE", () => putAtom.run(input.graphId, item["uid"], next, item["kind"], item["owner"], Buffer.from(refs), Buffer.from(payload), itemDigest, input.eventId));
                    yield* native("REWRITE", () => old ? updateHead.run(next, item["kind"], item["owner"], itemDigest, input.graphId, item["uid"]) : putHead.run(input.graphId, item["uid"], next, item["kind"], item["owner"], itemDigest));
                    produced.push(Object.freeze({ uid: item["uid"], revision: next, kind: item["kind"], owner: item["owner"], digest: itemDigest }));
                }
                const source = yield* json(input.source, "REWRITE");
                const consumedJson = yield* json(consumed, "REWRITE");
                const producedJson = yield* json(produced, "REWRITE");
                yield* native("REWRITE", () => database.prepare("INSERT INTO adaptive_events VALUES(?,?,?,?,?,?)").run(input.graphId, input.eventId, intent, Buffer.from(source), Buffer.from(consumedJson), Buffer.from(producedJson)));
                const outputHeads = Object.freeze(produced);
                return Object.freeze({ eventId: input.eventId, intentDigest: intent, source: input.source, consumed, produced: outputHeads, outputHeads });
            }));
        })),
        getEvent: (graphId: string, eventId: string) => effect(() => event("GET_EVENT", graphId, eventId)),
        events: (graphId: string) => effect(() => Either.gen(function* () { yield* graph("EVENTS", graphId); const found = yield* native("EVENTS", () => rows(database.prepare("SELECT event_id FROM adaptive_events WHERE graph_id=? ORDER BY rowid").all(graphId))); return yield* Either.all(found.map((item) => event("EVENTS", graphId, string(item["event_id"])))); })),
        acquireRuntimeLock: (graphId: string) => effect(() => Either.gen(function* () { if (!graphId)
            return yield* fail("ACQUIRE_RUNTIME_LOCK", "graph identity is invalid"); const token = randomUUID(); return yield* transaction("ACQUIRE_RUNTIME_LOCK", () => Either.gen(function* () { yield* graph("ACQUIRE_RUNTIME_LOCK", graphId); const holder = yield* native("ACQUIRE_RUNTIME_LOCK", () => row(database.prepare("SELECT token,pid FROM adaptive_runtime_locks WHERE graph_id=?").get(graphId))); if (holder && live(numeric(holder["pid"])))
            return yield* fail("ACQUIRE_RUNTIME_LOCK", "adaptive runtime lock is held by a live process"); if (holder)
            yield* native("ACQUIRE_RUNTIME_LOCK", () => database.prepare("DELETE FROM adaptive_runtime_locks WHERE graph_id=?").run(graphId)); yield* native("ACQUIRE_RUNTIME_LOCK", () => database.prepare("INSERT INTO adaptive_runtime_locks(graph_id,token,pid) VALUES(?,?,?)").run(graphId, token, process.pid)); return token; })); })),
        releaseRuntimeLock: (graphId: string, token: string) => effect(() => Either.gen(function* () { if (!graphId || !token)
            return yield* fail("RELEASE_RUNTIME_LOCK", "runtime lock identity is invalid"); return yield* transaction("RELEASE_RUNTIME_LOCK", () => Either.gen(function* () { yield* graph("RELEASE_RUNTIME_LOCK", graphId); const result = yield* native("RELEASE_RUNTIME_LOCK", () => database.prepare("DELETE FROM adaptive_runtime_locks WHERE graph_id=? AND token=?").run(graphId, token)); return numeric(result["changes"]) === 1 ? undefined : yield* fail("RELEASE_RUNTIME_LOCK", "adaptive runtime lock token does not match"); })); }))
    };
};
